cpp/oneapi/dal/backend/primitives/reduction/reduction_dpc.hpp

```diff
@@ -0,0 +1,70 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/backend/primitives/ndarray.hpp"
+#include "oneapi/dal/backend/primitives/reduction/functors.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+/// Reduces `input` rows inplace and stores results into `output`
+///
+/// @tparam Float    Floating-point type used to perform computations
+/// @tparam order    Input matrix data layout
+/// @tparam BinaryOp Type of binary operator functor
+/// @tparam UnaryOp  Type of unary operator functor
+///
+/// @param[in]  queue   The SYCL queue
+/// @param[in]  input   The [n x p] input dataset
+/// @param[out] output  The [n] results of reduction
+/// @param[in]  binary  The binary functor that reduces two values into one
+/// @param[in]  unary   The unary functor that will be performed element-wise before reduction
+/// @param[in]  deps    The vector of `sycl::event`s that represents list of dependencies
+template <typename Float, ndorder order, typename BinaryOp, typename UnaryOp>
```

Comment:
`BinaryOp` and `UnaryOp` shall have a clear contract. Now it seems that requirements for these classes expose the internal implementation of `reduce_by_rows`, because the requirement for such classes is to have `native` SYCL fields. This is DIP principle violation

Reply:
Discussed and fixed